src/Juvix/Backends/ArithmeticCircuit/Compilation/Types.hs

```diff
@@ -1,19 +1,48 @@
 module Juvix.Backends.ArithmeticCircuit.Compilation.Types where
 
-import qualified Data.Graph.Inductive as G
+import qualified Circuit
+import qualified Circuit.Expr as Expr
+import qualified Juvix.Backends.ArithmeticCircuit.Parameterisation as Par
+import qualified Juvix.Core.ErasedAnn as CoreErased
 import Juvix.Library
+import Numeric.Natural ()
 
-type Circuit = G.Gr NodeType WireType
+data PrimVal
+  = Element Par.F
+  | Boolean Bool
+  | FEInteger Int
+  | BinOp BinOp Term Term
+  | UnaryOp UnaryOp Term
+  | If Term Term Term
+  deriving (Show, Eq)
 
-data NodeType
-  = -- Addition gate.
-    Add
-  | -- Multiplication gate.
-    Mul
-  | -- Input value (by vector index).
-    InputValue Natural
-  | -- Output value (by vector index).
-    OutputValue Natural
-  deriving (Show, Eq, Generic)
+data BinOp
+  = Add
+  | Mul
+  | Sub
+  | Exp
+  | Eq
+  | And
+  | Or
+  deriving (Show, Eq)
 
-type WireType = ()
+data UnaryOp = Neg
+  deriving (Show, Eq)
+
+type Term = CoreErased.AnnTerm () PrimVal
```

Comment:
Is there a way to convert an `AnnTerm Ty Val` (from `Parameterisation.hs`)  to this type? We'll need that, since the type defined by the parameterisation is what the core type-checker will use.

Reply:
Probably? I just need a `* -> * -> *` datatype to implement the interface in Parameterisation.